src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,713 @@
+// src/main/c/NativeModule.cpp - native funcs for Scala NativeModule
+#include "hail/NativeModule.h"
+#include "hail/NativeObj.h"
+#include "hail/NativePtr.h"
+#include <assert.h>
+#include <jni.h>
+#include <dlfcn.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/time.h>
+#include <unistd.h>
+#include <atomic>
+#include <memory>
+#include <mutex>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#if 0
+#define D(fmt, ...) { \
+  char buf[1024]; \
+  sprintf(buf, fmt, ##__VA_ARGS__); \
+  fprintf(stderr, "DEBUG: %s,%d: %s", __FILE__, __LINE__, buf); \
+}
+#else
+#define D(fmt, ...) { }
+#endif
+
+namespace hail {
+
+namespace {
+
+// File-polling interval in usecs
+const int kFilePollMicrosecs = 50000;
+
+// Timeout for compile/link of a DLL
+const int kBuildTimeoutSecs = 120;
+
+// A quick-and-dirty way to get a hash of two strings, take 80bits,
+// and produce a 20byte string of hex digits.
+std::string hash_two_strings(const std::string& a, const std::string& b) {
+  auto hashA = std::hash<std::string>()(a);
+  auto hashB = std::hash<std::string>()(b);
+  hashA ^= (0x3ac5*hashB);
+  hashB &= 0xffff;
+  char buf[128];
+  char* out = buf;
+  for (int pos = 80; (pos -= 4) >= 0;) {
+    int64_t nibble = ((pos >= 64) ? (hashB >> (pos-64)) : (hashA >> pos)) & 0xf;
```

Comment:
We have already mixed all the bits of hashB into hashA (since 0x3ac5 is odd, when we multiplied
hashB*0x3ac5 we got at least one unshifted copy of all bits of A added into the results, so that
no bits of B were completely ignored).  It wouldn't necessarily matter if we discarded some bits
of B - to squeeze 128 bits down to 80 you have to lose *some* information - but the code doesn't
do that in the simple way of completely ignoring some bitfield (just because I'm slightly paranoid about whether the std::hash(std::string) is robust against doing that).

Reply:
Now changed to use 3 x 32bit hashes on a 32bit system, so that we still get a full 80 bits.